src/main/java/me/itzg/tsdbcassandra/services/DownsampleTrackingService.java

```diff
@@ -55,36 +53,75 @@ public DownsampleTrackingService(ReactiveCassandraTemplate cassandraTemplate,
         .hash();
     final int partition = Hashing.consistentHash(hashCode, properties.getPartitions());
+    final Instant normalizedTimeSlot = timestamp.with(timeSlotNormalizer);
 
-    return cassandraTemplate.update(
-        new PendingDownsampleSet()
-            .setPartition(partition)
-            .setTimeSlot(timestamp.with(timeSlotNormalizer))
-            .setTenant(tenant)
-            .setSeriesSet(seriesSet)
-            .setLastTouch(timestampProvider.now())
-    );
+    final String ingestingKey = PREFIX_INGESTING
+        + DELIM + partition
+        + DELIM + normalizedTimeSlot.getEpochSecond();
+
+    final String pendingKey = PREFIX_PENDING
+        + DELIM + partition
+        + DELIM + normalizedTimeSlot.getEpochSecond();
+    final String pendingValue = tenant
+        + DELIM + seriesSet;
+
+    return redisTemplate.opsForValue()
+        .set(ingestingKey, "", properties.getLastTouchDelay())
+        .and(
+            redisTemplate.opsForSet()
+            .add(pendingKey, pendingValue)
+        );
   }
 
-  public Flux<PendingDownsampleSet> retrieveReadyOnes(int partition) {
-    final Instant now = timestampProvider.now();
+  private static PendingDownsampleSet buildPending(String pendingKey, String pendingValue) {
+    final int splitValueAt = pendingValue.indexOf(DELIM);
 
-    return cassandraTemplate.select(
-        query(
-            where("partition").is(partition),
-            // make sure the whole slot has elapsed
-            where("timeSlot").lt(now.minus(properties.getTimeSlotWidth())),
-            // make sure slot isn't busy with updates
-            where("lastTouch").lt(now.minus(properties.getLastTouchDelay()))
+    return new PendingDownsampleSet()
+        .setTenant(pendingValue.substring(0, splitValueAt))
+        .setSeriesSet(pendingValue.substring(1+splitValueAt))
+        .setTimeSlot(Instant.ofEpochSecond(
+            Long.parseLong(pendingKey.substring(1 + pendingKey.lastIndexOf(DELIM)))
+        ));
+  }
+
+  public Flux<PendingDownsampleSet> retrieveReadyOnes(int partition) {
+    return redisTemplate
+        // scan over pending timeslots
+        .scan(
+            ScanOptions.scanOptions()
+                .match(PREFIX_PENDING + DELIM + partition + DELIM + "*")
+                .build()
         )
-            // for lastTouch filtering
-            .withAllowFiltering(),
-        PendingDownsampleSet.class
-    );
+        // expand each of those
+        .flatMap(pendingKey ->
+            // ...first see if the ingestion key
+            redisTemplate.hasKey(
+                PREFIX_INGESTING + pendingKey.substring(PREFIX_PENDING.length())
+            )
+                // ...has gone idle and been expired away
+                .filter(stillIngesting -> !stillIngesting)
+                // ...otherwise, expand by popping the downsample sets from that timeslot
+                .flatMapMany(stillIngesting -> popMany(pendingKey))
```

Comment:
so the pop is updating redis, before cassandra gets updated with the new data.  isn't that a problem when the process is restarted potentially losing any rollups in progress?

Reply:
also, stillIngesting seems like a bad name for the parameter to this lambda, because it isn't stillIngesting at this point right?

Reply:
Yeah, I was starting to worry about that too. I could have this instead do an `SSCAN` (set scan) and then an `SREM` to make that inter-processing more robust.

Reply:
> also, stillIngesting seems like a bad name for the parameter to this lambda, because it isn't stillIngesting at this point right?

Good point. The parameter isn't even used, so I'll name it something that emphasizes that stage.